[php-wasm-logger] Filter logs by severity in Logger and assign severity based on verbosity argument in CLIs

mho22 opened the pull request:
## Motivation for the change, related issues

The former intent of this pull request was to replace the console logs from the Xdebug bridge. 
Console logs were replaced by Logger logs and a filter by severity was added.

The CLIs verbosity argument is now related to its CLI LogVerbosity associated with a Logger LogSeverity. This setup will filter the related logs.

## Implementation details

The Playground CLI and Xdebug-bridge CLI now have a LogVerbosity variable : 

```typescript
const LogVerbosity = {
	Quiet: { name: 'quiet', severity: LogSeverity.Fatal },
	Normal: { name: 'normal', severity: LogSeverity.Info },
	Debug: { name: 'debug', severity: LogSeverity.Debug },
} as const;
```

The `yargs` `verbosity` option is now as follows : 

```typescript
.option('verbosity', {
	type: 'string',
	describe: 'Output logs',
	choices: Object.values(LogVerbosity).map(
		(verbosity) => verbosity.name
	),
	default: 'normal',
})
```

And the Logs are filtered based on the verbosity argument : 

```typescript
if (args.verbosity) {
    const severity = Object.values(LogVerbosity).find(
    	(v) => v.name === args.verbosity
    )!.severity;
    logger.filterBySeverity(severity);
}
```

The logger will then log the message based on if the severity level is equal or lower than the verbosity's severity level : 

```typescript
for (const handler of this.handlers) {
	if (logWithSeverity.severity.level <= this.severity.level) {
		handler(logWithSeverity, ...args);
	}
}
```

## Testing Instructions (or ideally a Blueprint)

### Normal mode

```
> nx run php-wasm-xdebug-bridge:dev

> node --no-warnings --experimental-wasm-stack-switching --experimental-wasm-jspi --loader=./packages/meta/src/node-es-module-loader/loader.mts ./packages/php-wasm/xdebug-bridge/src/cli.ts

Starting XDebug Bridge...
Connect Chrome DevTools to CDP at:
devtools://devtools/bundled/inspector.html?ws=localhost:9229

Chrome connected! Initializing Xdebug receiver...
XDebug receiver running on port 9003
Running a PHP script with Xdebug enabled...
```

### Quiet mode

```
> nx run php-wasm-xdebug-bridge:dev --verbosity quiet

> node --no-warnings --experimental-wasm-stack-switching --experimental-wasm-jspi --loader=./packages/meta/src/node-es-module-loader/loader.mts ./packages/php-wasm/xdebug-bridge/src/cli.ts --verbosity quiet

...silence...
```

### Debug mode

```
> nx run php-wasm-xdebug-bridge:dev --verbosity debug

> node --no-warnings --experimental-wasm-stack-switching --experimental-wasm-jspi --loader=./packages/meta/src/node-es-module-loader/loader.mts ./packages/php-wasm/xdebug-bridge/src/cli.ts --verbosity debug


Starting XDebug Bridge...
Connect Chrome DevTools to CDP at:
devtools://devtools/bundled/inspector.html?ws=localhost:9229

[CDP][received] {"id":1,"method":"Network.setCacheDisabled","params":{"cacheDisabled":true}}
[CDP][received] {"id":2,"method":"Network.enable","params":{"maxPostDataSize":65536,"reportDirectSocketTraffic":true}}
[CDP][received] {"id":3,"method":"Network.setAttachDebugStack","params":{"enabled":true}}
[CDP][received] {"id":4,"method":"Page.enable","params":{}}
[CDP][received] {"id":5,"method":"Page.getResourceTree","params":{}}
[CDP][received] {"id":6,"method":"Runtime.enable","params":{}}
[CDP][received] {"id":7,"method":"DOM.enable","params":{}}
[CDP][received] {"id":8,"method":"CSS.enable","params":{}}
[CDP][received] {"id":9,"method":"Debugger.enable","params":{"maxScriptsCacheSize":10000000}}
[CDP][received] {"id":10,"method":"Debugger.setPauseOnExceptions","params":{"state":"none"}}
[CDP][received] {"id":11,"method":"Debugger.setAsyncCallStackDepth","params":{"maxDepth":32}}
[CDP][received] {"id":12,"method":"Overlay.enable","params":{}}
[CDP][received] {"id":13,"method":"Overlay.setShowViewportSizeOnResize","params":{"show":true}}
[CDP][received] {"id":14,"method":"Animation.enable","params":{}}
[CDP][received] {"id":15,"method":"Autofill.enable","params":{}}
[CDP][received] {"id":16,"method":"Autofill.setAddresses","params":{"addresses":[]}}
[CDP][received] {"id":17,"method":"Profiler.enable","params":{}}
[CDP][received] {"id":18,"method":"Log.enable","params":{}}
[CDP][received] {"id":19,"method":"Log.startViolationsReport","params":{"config":[{"name":"longTask","threshold":200},{"name":"longLayout","threshold":30},{"name":"blockedEvent","threshold":100},{"name":"blockedParser","threshold":-1},{"name":"handler","threshold":150},{"name":"recurringHandler","threshold":50},{"name":"discouragedAPIUse","threshold":-1}]}}
[CDP][received] {"id":20,"method":"Emulation.setEmulatedMedia","params":{"media":"","features":[{"name":"color-gamut","value":""},{"name":"prefers-color-scheme","value":""},{"name":"forced-colors","value":""},{"name":"prefers-contrast","value":""},{"name":"prefers-reduced-data","value":""},{"name":"prefers-reduced-motion","value":""},{"name":"prefers-reduced-transparency","value":""}]}}
[CDP][received] {"id":21,"method":"Emulation.setEmulatedVisionDeficiency","params":{"type":"none"}}
[CDP][received] {"id":22,"method":"Audits.enable","params":{}}
[CDP][received] {"id":23,"method":"ServiceWorker.enable","params":{}}
[CDP][received] {"id":24,"method":"Inspector.enable","params":{}}
[CDP][received] {"id":25,"method":"Target.setAutoAttach","params":{"autoAttach":true,"waitForDebuggerOnStart":true,"flatten":true}}
[CDP][received] {"id":26,"method":"Target.setDiscoverTargets","params":{"discover":true}}
[CDP][received] {"id":27,"method":"Target.setRemoteLocations","params":{"locations":[{"host":"localhost","port":9229}]}}
[CDP][received] {"id":28,"method":"Runtime.addBinding","params":{"name":"__chromium_devtools_metrics_reporter","executionContextName":"DevTools Performance Metrics"}}
[CDP][received] {"id":29,"method":"Network.clearAcceptedEncodingsOverride","params":{}}
[CDP][received] {"id":30,"method":"Debugger.setBlackboxPatterns","params":{"patterns":["/node_modules/|^node:"],"skipAnonymous":false}}
[CDP][received] {"id":31,"method":"Page.getNavigationHistory","params":{}}
[CDP][received] {"id":32,"method":"DOMDebugger.setBreakOnCSPViolation","params":{"violationTypes":[]}}
[CDP][received] {"id":33,"method":"Debugger.setBreakpointByUrl","params":{"lineNumber":11,"url":"http://[::1]:5173/@fs/Users/mho/Work/Sandbox/Web/js/php-wasm/wordpress-playground/dist/packages/php-wasm/logger/index.js","columnNumber":0,"condition":""}}
[CDP][received] {"id":34,"method":"Debugger.setBreakpointByUrl","params":{"lineNumber":42,"url":"https://sparse.test:5173/@fs/Users/mho/Work/Sandbox/Web/js/php-wasm/wordpress-playground/dist/packages/php-wasm/web/php/jspi/php_8_2.js","columnNumber":68886,"condition":""}}
[CDP][received] {"id":35,"method":"Debugger.setBreakpointByUrl","params":{"lineNumber":4480,"url":"debugger:///VM902 client-CdekbSNg.js","columnNumber":0,"condition":""}}
[CDP][received] {"id":36,"method":"Debugger.setBreakpointByUrl","params":{"lineNumber":2,"url":"file:///Users/mho/Work/Projects/Development/Web/Professional/xdebug/template/php/xdebug.php","columnNumber":0,"condition":""}}
[CDP][received] {"id":37,"method":"Debugger.setBreakpointByUrl","params":{"lineNumber":2,"url":"file://xdebug.php","columnNumber":0,"condition":""}}
[CDP][received] {"id":38,"method":"Debugger.setBreakpointByUrl","params":{"lineNumber":2,"url":"file:///wordpress/xdebug.php","columnNumber":0,"condition":""}}
[CDP][received] {"id":39,"method":"Runtime.runIfWaitingForDebugger","params":{}}
[CDP][received] {"id":40,"method":"Page.setAdBlockingEnabled","params":{"enabled":false}}
[CDP][received] {"id":41,"method":"Emulation.setFocusEmulationEnabled","params":{"enabled":false}}
[CDP][received] {"id":42,"method":"Page.startScreencast","params":{"format":"jpeg","quality":80,"maxWidth":272,"maxHeight":1778}}
Chrome connected! Initializing Xdebug receiver...
XDebug receiver running on port 9003
Running a PHP script with Xdebug enabled...
```

mho22 commented:
I will try to add tests before marking it as ready for review.

adamziel commented:
Maybe `--verbosity=quiet|normal|debug` or so? That way we won't run into `./run-bridge --quiet --verbose` conundrum. Also, we can always call `logger.log()` and such – output filtering can be done at the [logToConsole](https://github.com/WordPress/wordpress-playground/blob/78a353e9b8a82896ea43f1193ac26332a10ae970/packages/php-wasm/logger/src/lib/handlers/log-to-console.ts#L8) log handler level (with some refactoring). Then, the CLI package could follow that pattern.

mho22 commented:
@adamziel This one was a quite big refactoring. I will need a review on this. A little summary of what I had in mind when I coded these : 

1. I set a LogVerbosity  option like you suggested. Directly inside the Logger file. 3 options : `quiet`, `normal` and `debug`.

2. The default one is `debug` since `debug` accepts every severity, like before the refactoring.

3. When calling `filterByVerbosity` : 

- if you choose `quiet` no severity will be printed in the console
- if you choose `normal` all severities except `debug` will be printed in the console.
- if you choose `debug` all severities  will be printed in the console.

4. I implemented the `--verbosity` option in : 

- The Xdebug Bridge CLI
- The Playground CLI

5. I implemented tests in : 

- Logger
- Xdebug Bridge
- Playground CLI



I therefore have several questions : 

1. The `normal` verbosity should silence more than only the `debug` severity ?
2. Should I add a new `--verbosity` option in the PHP Wasm CLI too ?
3. Should I apply these filters to every LogHandlers or only the `logToConsole` one ?

adamziel commented on packages/php-wasm/logger/src/lib/logger.ts:
```diff
@@ -62,14 +89,39 @@ export class Logger extends EventTarget {
 	public logMessage(log: Log, ...args: any[]): void {
 		for (const handler of this.handlers) {
-			handler(log, ...args);
+			const isConsole = handler.name === 'logToConsole';
+			const isSeverityAllowed = log.severity
+				? this.filters.includes(log.severity)
+				: !!this.filters.length;
+
+			if (!isConsole || isSeverityAllowed) {
+				handler(log, ...args);
+			}
```
I'm confused – what's the reasoning here? Do we need special casing?

mho22 replied:
If you're asking for the `isConsole` line : in my last comment I wanted to know if your suggestion with `--verbosity`  only concerned `logToConsole`. If not I can remove that line. 

If you're asking for the `isSeverityAllowed` line, the severity of a log can be a LogSeverity type but also `undefined`. And if it is undefined, it takes the `Info` severity in `log-to-memory`. That is why I had to use a ternary operator. 

I would suggest to replace the undefined severity by a `Log` `LogServerity` and in `log-to-memory.ts` replace:  

```diff
- log.severity ?? LogSeverity.Info,
+ log.severity == LogSeverity.Log ? LogSeverity.Info : log.severity
```

adamziel replied:
> in my last comment I wanted to know if your suggestion with --verbosity only concerned logToConsole.

Ah, good question! Let's target all handlers. If we want to restrict the console output while still collecting all the logs somewhere, we can do that at the `logToConsole` handler level.

mho22 replied:
Done.

adamziel commented on packages/php-wasm/logger/src/lib/logger.ts:
```diff
@@ -62,14 +89,39 @@ export class Logger extends EventTarget {
+	 * @param verbosity LogVerbosity
+	 */
+	public filterByVerbosity(verbosity: LogVerbosity): void {
+		if (verbosity === LogVerbosity.Quiet) {
```
We could associate verbosities with numbers and compare `messageLevel >= loggerVerbosity` (or `<=`? let's check how other libraries approach this)

mho22 replied:
On it.

mho22 replied:
Done.

adamziel commented on packages/playground/cli/src/run-cli.ts:
```diff
@@ -142,10 +142,11 @@ export async function parseOptionsAndRunCLI() {
 				default: false,
 			})
-			.option('quiet', {
+			.option('verbosity', {
```
Let's still accept the `--quiet` option and translate it into verbosity to avoid breaking BC. We can just hide it from the help message (`hidden: true` I think?)

mho22 replied:
Done.

adamziel commented:
Things are looking good! I haven't tested it yet at this stage, only left a few comments – [this](https://github.com/WordPress/wordpress-playground/pull/2436#discussion_r2256998111) one seems the most important.

adamziel commented:
> The normal verbosity should silence more than only the debug severity ?

I'd say normal includes all the log levels except `debug`.

> Should I add a new --verbosity option in the PHP Wasm CLI too ?

Potentially? But not in this PR, it already covers a lot of ground.

mho22 commented:
> I'd say normal includes all the log levels except debug.

Done.

> Potentially? But not in this PR, it already covers a lot of ground.

Ok! What do you think about using `yargs` in `php-wasm/cli/main.ts` and renaming the file to `run-cli.ts` to maintain consistency across the different CLIs? Let's talk about that in the next pull request maybe.

adamziel commented:
> Ok! What do you think about using yargs in php-wasm/cli/main.ts and renaming the file to run-cli.ts to maintain consistency across the different CLIs? Let's talk about that in the next pull request maybe.

Sounds great for a follow-up PR!

adamziel commented on packages/playground/cli/src/blueprints-v1/blueprints-v1-handler.ts:
```diff
@@ -96,7 +96,7 @@ export class BlueprintsV1Handler {
 				progressReached100 = percentProgress === 100;
 
-				if (!this.args.quiet) {
+				if (this.args.verbosity !== LogVerbosity.Quiet) {
```
Let's check it in `this.writeProgressUpdate` instead of before every call

mho22 replied:
done.

adamziel commented:
I'll test this and follow up tomorrow

adamziel commented on packages/php-wasm/logger/src/lib/logger.ts:
```diff
@@ -12,27 +12,54 @@ export { errorLogPath } from './collectors/collect-php-logs';
+	Normal: 'normal',
+	Quiet: 'quiet',
+	Debug: 'debug',
+} as const;
```
Let's not distinguish between verbosity and severity at the logger level. One concept is enough. We can decide what each verbosity means in the CLI args parser when we let the user decide. All internal methods could use severities for filtering.

mho22 replied:
done.

adamziel commented on packages/php-wasm/logger/src/lib/logger.ts:
```diff
@@ -12,27 +12,54 @@ export { errorLogPath } from './collectors/collect-php-logs';
+	Error: { name: 'error', level: 1 },
+	Fatal: { name: 'fatal', level: 1 },
+	Debug: { name: 'debug', level: 2 },
+} as const;
```
Let's give each severity a unique level, e.g.

```ts
export const LogSeverity = {
	Fatal: { name: 'fatal', level: 0 },
	Error: { name: 'error', level: 1 },
	Warn: { name: 'warn', level: 2 },
	Log: { name: 'log', level: 3 },
	Info: { name: 'info', level: 4 },
	Debug: { name: 'debug', level: 5 },
} as const;
```

mho22 replied:
done.

adamziel commented:
I've left two notes about the code design, and my only other note would be to adjust the PR title and description to reflect the larger logging refactor.

mho22 commented:
I added the corrections you asked and modified the bridge to respect the logic from the playground : 

- Removed the LogVerbosity from Logger
- Added unique level to LogSeverities
- Replaced `filterByVerbosity` with `filterBySeverity`
- Added a LogVerbosity const in CLIs
- Filtered logs by severity in CLIs
- Removed verbosity option from `xdebug-bridge/start-bridge.ts` ( if bridge is used in code, logger also can be used ).
- Add tests for the verbosity argument in `xdebug-bridge/run-cli.ts`
- Add tests for logs in `xdebug-bridge/start-bridge.ts`

I will now adjust the pull request title and description.

- I set the `quiet` verbosity to be of severity `Fatal` (its level is 0), so even if the verbosity is set to `quiet` it should log the fatal error.

brandonpayton commented on packages/php-wasm/logger/project.json:
```diff
@@ -64,15 +64,15 @@
-				"tsc -p packages/php-wasm/logger/tsconfig.spec.json --noEmit"
-			]
+		},
+		"typecheck": {
```
good catch :)

brandonpayton commented on packages/php-wasm/logger/src/lib/logger.ts:
```diff
@@ -12,27 +12,43 @@ export { errorLogPath } from './collectors/collect-php-logs';
+	Debug: { name: 'debug', level: 5 },
+} as const;
+
+export type LogSeverity = (typeof LogSeverity)[keyof typeof LogSeverity];
```
Nice

brandonpayton commented on packages/php-wasm/logger/src/lib/logger.ts:
```diff
@@ -12,27 +12,43 @@ export { errorLogPath } from './collectors/collect-php-logs';
+	JS: 'JavaScript',
+} as const;
+
+export type LogPrefix = (typeof LogPrefix)[keyof typeof LogPrefix];
```
This is great!

brandonpayton commented on packages/php-wasm/logger/src/lib/logger.ts:
```diff
@@ -12,27 +12,43 @@ export { errorLogPath } from './collectors/collect-php-logs';
 export class Logger extends EventTarget {
 	public readonly fatalErrorEvent = 'playground-fatal-error';
-	private readonly handlers: LogHandler[];
+	private handlers: LogHandler[];
```
Why did we remove the `readonly` marker? AFAICT, there isn't a change in how we privately treat `handlers` in this class.

mho22 replied:
You're right, I'm reverting the change.

mho22 replied:
done.

brandonpayton commented on packages/php-wasm/logger/src/lib/logger.ts:
```diff
@@ -62,17 +78,32 @@ export class Logger extends EventTarget {
+	 * Filter message based on severity
+	 * @param severity LogSeverity
+	 */
+	public filterBySeverity(severity: LogSeverity): void {
```
When I read this method name elsewhere, its name sounded like a verb, like it was actually doing the filtering. Could we clarify the name a bit so it sounds more like a setter, even something as verbose as `setSeverityFilterLevel`?

mho22 replied:
Good idea!

mho22 replied:
done.

brandonpayton commented on packages/php-wasm/xdebug-bridge/README.md:
```diff
@@ -37,32 +41,45 @@ await server.start();
+-   `port`: Xdebug port to listen on (default: 9003)
+-   `host`: Xdebug host to bind to (default: 'localhost')
+-   `php-root`: Path to PHP root directory (default: './')
+-   `verbosity`: Output logs and progress messages (default: 'normal')
```
Can we list possible `severity` values here and below in the API section?

mho22 replied:
Yes! On it.

mho22 replied:
done.

brandonpayton commented on packages/php-wasm/xdebug-bridge/src/lib/xdebug-cdp-bridge.ts:
```diff
@@ -136,6 +136,11 @@ export class XdebugCDPBridge {
 		});
 	}
 
+	stop() {
```
Is this change intentionally part of this PR? No complaint here. Just confirming it isn't accidentally slipping in.

mho22 replied:
Yes. It was needed for the tests and added in another pull request.

brandonpayton commented on packages/playground/cli/README.md:
```diff
@@ -93,7 +93,7 @@ The `server` command supports the following optional arguments:
 -   `--skip-wordpress-setup`: Do not download or install WordPress. Useful if you are mounting a full WordPress directory.
 -   `--skip-sqlite-setup`: Do not set up the SQLite database integration.
--   `--quiet`: Do not output logs and progress messages.
+-   `--verbosity`: Output logs and progress messages.
```
It seems unfortunate to explicitly document these values everywhere, but it would help readers. What do you think about mentioning the possible values here?

mho22 replied:
On it!

mho22 replied:
done.

brandonpayton commented on packages/playground/cli/src/run-cli.ts:
```diff
@@ -142,10 +150,20 @@ export async function parseOptionsAndRunCLI() {
 				type: 'boolean',
 				default: false,
 			})
+			// Hidden - Deprecated in favor of verbosity
```
👍

brandonpayton commented on packages/playground/cli/src/run-cli.ts:
```diff
@@ -357,9 +376,26 @@ export async function runCLI(args: RunCLIArgs): Promise<RunCLIServer> {
+
+	// Promote "debug" flag to verbosity but keep args.debug around – the
+	// program behavior may change in more ways than just logging verbosity
+	// when debug mode is enabled, e.g. error objects may carry additional details.
```
Nice explanation.

brandonpayton commented on packages/php-wasm/logger/src/test/logger.spec.ts:
```diff
@@ -1,24 +1,97 @@
+		logger.debug('debug');
+		const logs = logger.getLogs();
+		expect(logs.length).toBe(5);
+		expect(output).toEqual(['log', 'info', 'warn', 'error', 'debug']);
```
These are nice, readable tests.

brandonpayton reviewed (approved):
@mho22, these changes look great to me. We've needed something like this for a while, and I think we'll be able to build on this to further improve/implement other code like the Playground CLI's --experimental-trace flag.

I left some questions and comments. Most of the comments simple "I like this" kinds of comments, but there are others too.

brandonpayton commented:
@mho22, if you believe this PR is complete and have no other concerns, please feel free to merge this.

mho22 commented:
I updated the code with your suggestions. The tests passed. I guess this can be merged.
